protocol/client.go

```diff
@@ -0,0 +1,193 @@
+		CurrentEpoch: 0,
+		SavedSTR:     savedSTR,
+		isUseTBs:     isUseTBs,
+		signKey:      signKey,
+	}
+}
+
+func (cs *ConiksClient) Verify(msg *Response, uname string, key []byte) {
+	switch msg.DirectoryResponse.(type) {
+	case *DirectoryProof:
+		cs.verifyDirectoryProof(
+			msg.DirectoryResponse.(*DirectoryProof), uname, key)
+	case *DirectoryProofs:
+		cs.verifyDirectoryProofs(
+			msg.DirectoryResponse.(*DirectoryProofs), uname, key)
+	}
+}
+
+func (cs *ConiksClient) verifyDirectoryProof(df *DirectoryProof, uname string, key []byte) {
+	ap := df.AP
+	str := df.STR
+	tb := df.TB
+
+	if cs.VerificationResult = cs.verifySTR(str); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.ProofType, cs.VerificationResult = cs.verifyAuthPath(uname, key,
+		ap, str); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.VerificationResult = cs.verifyReturnedPromise(tb, str, ap); cs.VerificationResult != Passed {
+		return
+	}
+
+	if cs.VerificationResult = cs.verifyIssuedPromises(str, ap); cs.VerificationResult != Passed {
+		return
+	}
+
+	cs.VerificationResult = Passed
+
+	// re-assign new state for the client
+	if str.Epoch > cs.CurrentEpoch {
+		cs.CurrentEpoch++
+		cs.SavedSTR = str.Signature
+	}
+}
+
+func (cs *ConiksClient) verifyDirectoryProofs(dfs *DirectoryProofs, uname string, key []byte) {
+	// TODO: fix me
+	cs.VerificationResult = Passed
+}
+
+func (cs *ConiksClient) verifyAuthPath(uname string, key []byte,
+	ap *m.AuthenticationPath,
+	str *m.SignedTreeRoot) (ProofType, ErrorCode) {
+	proofType := ProofOfAbsence
+
+	// verify VRF Index
+	vrfKey := vrf.PublicKey(str.Policies.VrfPublicKey)
+	if !vrfKey.Verify([]byte(uname), ap.LookupIndex, ap.VrfProof) {
+		return InvalidProof, ErrorBadVRFProof
+	}
+
+	if bytes.Equal(ap.LookupIndex, ap.Leaf.Index) { /* proof of inclusion */
+		// verify name-to-key binding
+		// key is nil when the user does lookup for the first time
+		if key != nil && !ap.VerifyBinding(key) {
+			return InvalidProof, ErrorBadBinding
+		}
+		// verify commitment
+		if !ap.Leaf.Commitment.Verify([]byte(uname), ap.Leaf.Value) {
+			return InvalidProof, ErrorBadCommitment
+		}
+		proofType = ProofOfInclusion
+	}
+
+	// verify auth path
+	if !ap.Verify(str.TreeHash) {
+		return InvalidProof, ErrorBadAuthPath
+	}
+
+	return proofType, Passed
+}
+
+func (cs *ConiksClient) verifySTR(str *m.SignedTreeRoot) ErrorCode {
+	// verify STR's signature
+	if !cs.signKey.Verify(str.Serialize(), str.Signature) {
+		return ErrorBadSignature
+	}
+
+	// verify hash chain
+	if (str.Epoch == cs.CurrentEpoch && bytes.Equal(cs.SavedSTR, str.Signature)) ||
+		(str.Epoch == cs.CurrentEpoch+1 && str.VerifyHashChain(cs.SavedSTR)) {
+		return Passed
+	}
+	return ErrorBadSTR
+}
+
+// verifyReturnedPromise verifies the returned TB
+// included in the directory's response.
+func (cs *ConiksClient) verifyReturnedPromise(tb *m.TemporaryBinding,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) ErrorCode {
+	if cs.isUseTBs && tb != nil {
```

Comment:
> Forgive the naive question: why does the client have control of whether to use TBs? Shouldn't useTB be part of the policies of the current epoch and the client just react to that?

As far as I understood it, @c633 didn't intend to include the flag for client to decide if he wants to support/verify TBs but just to store all the data in the struct necessary for the actual client to verify everything (including TBs if necessary). 

I think the confusion stems from the fact that it looks like the client could just decide how to handle TBs; which he shouldn't (but always could, as the server isn't able to force the client to act different).

Maybe, instead the constructor, `NewConiksClient`,  should take the policy instead of a boolean flag to make this semantically clearer? The client could then use the policy to decide if he needs to verify TBs or not. Does this make sense?